File: 07-security/01-authn-authz-and-idor/golang/idor_enumeration.go
// Layer 7 · Topic 1 — IDOR as a missing authorization check (Go).
//
// One command, no arguments, no deps: `go run idor_enumeration.go`.
// Seeds 1,500 invoices across three interleaved tenants, logs in as alice,
// fires 1,000 enumeration requests at three handler variants, and counts the
// wrong-owner rows leaked per 1,000.
//
// Go's contribution here is ergonomic, not a safety property: with no ORM
// doing implicit fetches, the SELECT is written out in front of you, so the
// missing `AND owner_id = $2` is visible in the diff. invoiceID is an int64
// whether or not it is *yours* -- static typing does not help at all. That is
// the entire difference from the other runtimes, and it is a readability win,
// not a security one.
//
// What to look for: vulnerable leaks ~667/1000; both fixes leak 0 and return
// 404 (not 403) on a wrong-owner id.
package main

import "fmt"

const (
	nInvoices  = 1500
	caller     = 1 // alice
	nRequests  = 1000
)

type invoice struct {
	id      int
	ownerID int
	amount  int
}

type store struct {
	rows        map[int]invoice
	requestUser int // 0 == unset; the RLS "session variable"
}

func newStore() *store {
	s := &store{rows: make(map[int]invoice, nInvoices), requestUser: 0}
	for i := 1; i <= nInvoices; i++ {
		s.rows[i] = invoice{id: i, ownerID: ((i - 1) % 3) + 1, amount: i * 100}
	}
	return s
}

// vulnerable: db.QueryRow("SELECT ... WHERE id=$1", id) -- no caller.
func (s *store) get(id int) (invoice, bool) { r, ok := s.rows[id]; return r, ok }

// query-layer fix: ... WHERE id=$1 AND owner_id=$2
func (s *store) getScoped(id, owner int) (invoice, bool) {
	r, ok := s.rows[id]
	if ok && r.ownerID == owner {
		return r, true
	}
	return invoice{}, false
}

// data-layer fix (RLS analogue): policy enforced below the handler.
func (s *store) getRLS(id int) (invoice, bool) {
	r, ok := s.rows[id]
	if !ok || s.requestUser == 0 || r.ownerID != s.requestUser {
		return invoice{}, false
	}
	return r, true
}

type handler func(s *store, caller, id int) (int, *invoice)

func vulnerable(s *store, caller, id int) (int, *invoice) {
	if r, ok := s.get(id); ok {
		return 200, &r
	}
	return 404, nil
}
func fixedQuery(s *store, caller, id int) (int, *invoice) {
	if r, ok := s.getScoped(id, caller); ok {
		return 200, &r
	}
	return 404, nil
}
func fixedRLS(s *store, caller, id int) (int, *invoice) {
	s.requestUser = caller // SET LOCAL app.current_user
	if r, ok := s.getRLS(id); ok {
		return 200, &r
	}
	return 404, nil
}

// Deterministic LCG in uint32 space -> reproducible, measured numbers.
func enumIDs(n int) []int {
	ids := make([]int, n)
	var x uint32 = 1
	for i := 0; i < n; i++ {
		x = 1103515245*x + 12345
		ids[i] = 1 + int(x%nInvoices)
	}
	return ids
}

func runVariant(h handler, label string) {
	s := newStore()
	leaked, own, notFound := 0, 0, 0
	for _, id := range enumIDs(nRequests) {
		status, row := h(s, caller, id)
		if status == 200 {
			if row.ownerID != caller {
				leaked++
			} else {
				own++
			}
		} else {
			notFound++
		}
	}
	wrongStatus, _ := h(newStore(), caller, 2) // id 2 is bob's
	fmt.Printf("  %-14s leaked=%4d/%d   own=%4d  not_found=%4d   wrong-owner id -> HTTP %d\n",
		label, leaked, nRequests, own, notFound, wrongStatus)
}

func main() {
	fmt.Println("Layer 7 · Topic 1 — IDOR enumeration (logged in as alice, id=1)")
	fmt.Printf("seed: %d invoices, 3 tenants interleaved, alice owns %d\n", nInvoices, nInvoices/3)
	fmt.Printf("attack: %d requests, ids uniform over 1..%d\n\n", nRequests, nInvoices)
	fmt.Println("  handler        wrong-owner rows leaked per 1,000 requests")
	runVariant(vulnerable, "vulnerable")
	runVariant(fixedQuery, "fixed_query")
	runVariant(fixedRLS, "fixed_rls")
	fmt.Println("\nRead: the SELECT is visible, so the missing owner filter is " +
		"visible too -- but nothing forces you to write it. The fix that survives " +
		"a team is the one the next author cannot omit: the data layer (RLS).")
}

File: 07-security/01-authn-authz-and-idor/java/IdorEnumeration.java
// Layer 7 · Topic 1 — IDOR as a missing authorization check (Java).
//
// One command, no arguments, JDK only:
//   javac IdorEnumeration.java && java IdorEnumeration
//
// Java is the one runtime in this topic with a DECLARATIVE authorization
// model: Spring Security's @PreAuthorize("@invoiceGuard.owns(#id, auth)")
// runs before the method body and is greppable at the signature. That is a
// genuine improvement AND the sharpest illustration of its limit: the check
// is opt-in per method, so an UNANNOTATED new endpoint is wide open, and
// Spring Data's findById(id) underneath is still a primary-key lookup. A
// declarative check you can forget to declare has the same failure mode as
// an `if` you can forget to write. We model exactly that: `vulnerable` is the
// endpoint whose @PreAuthorize was never added.
//
// What to look for: vulnerable leaks ~667/1000; both fixes leak 0 and return
// 404 (not 403) on a wrong-owner id.
import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;

public class IdorEnumeration {
    static final int N_INVOICES = 1500;
    static final int CALLER = 1; // alice
    static final int N_REQUESTS = 1000;

    record Invoice(int id, int ownerId, int amount) {}

    static final class Store {
        final Map<Integer, Invoice> rows = new HashMap<>();
        int requestUser = 0; // the RLS "session variable"
        Store() {
            for (int i = 1; i <= N_INVOICES; i++)
                rows.put(i, new Invoice(i, ((i - 1) % 3) + 1, i * 100));
        }
        Invoice get(int id) { return rows.get(id); }            // findById(id)
        Invoice getScoped(int id, int owner) {                  // WHERE id AND owner_id
            Invoice r = rows.get(id);
            return (r != null && r.ownerId() == owner) ? r : null;
        }
        Invoice getRls(int id) {                                // policy below the handler
            Invoice r = rows.get(id);
            if (r == null || requestUser == 0 || r.ownerId() != requestUser) return null;
            return r;
        }
    }

    // A handler returns {status, ownerIdOrMinus1}.
    interface Handler { int[] handle(Store s, int caller, int id); }

    static int[] vulnerable(Store s, int caller, int id) {
        Invoice r = s.get(id);                                  // @PreAuthorize never added
        return r != null ? new int[]{200, r.ownerId()} : new int[]{404, -1};
    }
    static int[] fixedQuery(Store s, int caller, int id) {
        Invoice r = s.getScoped(id, caller);
        return r != null ? new int[]{200, r.ownerId()} : new int[]{404, -1};
    }
    static int[] fixedRls(Store s, int caller, int id) {
        s.requestUser = caller;                                 // SET LOCAL app.current_user
        Invoice r = s.getRls(id);
        return r != null ? new int[]{200, r.ownerId()} : new int[]{404, -1};
    }

    static int[] enumIds(int n) {
        int[] ids = new int[n];
        int x = 1; // LCG in 32-bit space; deterministic, measured numbers
        for (int i = 0; i < n; i++) {
            x = 1103515245 * x + 12345;
            ids[i] = 1 + Math.floorMod(x, N_INVOICES);
        }
        return ids;
    }

    static void run(Handler h, String label) {
        Store s = new Store();
        int leaked = 0, own = 0, notFound = 0;
        for (int id : enumIds(N_REQUESTS)) {
            int[] resp = h.handle(s, CALLER, id);
            if (resp[0] == 200) { if (resp[1] != CALLER) leaked++; else own++; }
            else notFound++;
        }
        int wrongStatus = h.handle(new Store(), CALLER, 2)[0]; // id 2 is bob's
        System.out.printf("  %-14s leaked=%4d/%d   own=%4d  not_found=%4d   wrong-owner id -> HTTP %d%n",
                label, leaked, N_REQUESTS, own, notFound, wrongStatus);
    }

    public static void main(String[] args) {
        System.out.println("Layer 7 · Topic 1 — IDOR enumeration (logged in as alice, id=1)");
        System.out.printf("seed: %d invoices, 3 tenants interleaved, alice owns %d%n", N_INVOICES, N_INVOICES / 3);
        System.out.printf("attack: %d requests, ids uniform over 1..%d%n%n", N_REQUESTS, N_INVOICES);
        System.out.println("  handler        wrong-owner rows leaked per 1,000 requests");
        run(IdorEnumeration::vulnerable, "vulnerable");
        run(IdorEnumeration::fixedQuery, "fixed_query");
        run(IdorEnumeration::fixedRls, "fixed_rls");
        System.out.println("\nRead: @PreAuthorize makes the check visible at the signature, but " +
                "it is opt-in -- the unannotated endpoint is the leak. The data-layer control " +
                "(RLS) is the one a missing annotation cannot switch off.");
    }
}
